### ISSUE TITLE
Correct setup of MongoDB to create databases 'on the fly'

### ISSUE THREAD

**Issue by dww100** [radical-collaboration/GB-SC18#2]
I have been playing with the getting started script for ensemblemd. If I set the MongoDB URL environment variable using:
`export RADICAL_PILOT_DBURL='mongodb://user:password@crick.chem.ucl.ac.uk:27017/testdb'`
and comment out the database_url entry in the ResourceHandle then I would expect the testdb databse to be created on the fly as suggested by:
http://radicalpilot.readthedocs.io/en/latest/installation.html#mongodb-service
Instead authentication fails giving the following error:
> radical.pilot.exceptions.PilotException: Couldn't create new session (database URL 'mongodb://user:password@crick.chem.ucl.ac.uk:27017/testdb' incorrect?): command SON([('authenticate', 1), ('user', u'user'), ('nonce', u'ef93882e62db1210'), ('key', u'f7d6f92c6e7f4923354b7f588487fdc4')]) on namespace testdb.$cmd failed: auth failed

If I change testdb to admin then the process proceeds correctly but the run uses the admin database.

If I log into the mongodb shell I can happily create new databases with the user incorporated in the URL.

Any idea what changes need to be made to our setup?

**Comment by vivek-bala** [comment]
Hey Dave,

Which branch of Radical pilot and Ensemble toolkit are you using? You can also get the versions by doing ```radicalpilot-version``` and ```ensemblemd-version```.

``` 
(database URL 'mongodb://user:password@crick.chem.ucl.ac.uk:27017/testdb' incorrect?)
```
Just to be sure. After setting up a mongodb instance (on mlab for example) you will need to add a user and password for the user. Those credentials needs to be substituted into this URL. Ignore if that is already done and scrapped off since ticket is public.

**Comment by dww100** [comment]
The username and password are indeed placeholders. In reality we have used a real username and password we have previously created. We have also checked we can make new databases with this used on the mongodb shell. I think it would be helpful if the need to do these things was documented somewhere as the roles we had to give the user were not default (or necessarily intuitive). I'll happily do that once we have things sorted out.

I'm using:
`radicalpilot` version 0.44
`ensemblemd` version 0.4.6

**Comment by vivek-bala** [comment]
That is a bit weird. Just to check: I tried out the same example and used the environment variable for the mongodb url, seems to have worked for me. 

Two questions:

* Were there any changes to the ```get_started.py``` script? I am assuming you already commented out the dburl parameter in the ResourceHandle object in the script
* Does the default dburl in the script work?

**Comment by dww100** [comment]
* I haven't altered the `get_started.py` script other than to experiment with changing the hard coded database_url (I've tried commenting out and setting a new hard coded address).

* The default database_url is working fine

I'm using the conda version of ensemble toolkit by the way

**Comment by vivek-bala** [comment]
> * I haven't altered the ```get_started.py``` script other than to experiment with changing the hard coded database_url (I've tried commenting out and setting a new hard coded address).

Hmmm. Strange. Is it possible for me to get access to the same mongo db?

> I'm using the conda version of ensemble toolkit by the way

Ah ok. I will test in the same environment then.

**Comment by dww100** [comment]
Sure - the mongodb instance is at:
Host: crick.chem.ucl.ac.uk
Port: 27017
Username: radical
Password: fg*2GT3^eB

@sjzasada let me know if either a) this is and issue or b) there is more information needed

**Comment by vivek-bala** [comment]
I get an authentication error. I assume the db name is 'testdb' based on the error log in the ticket. Can you confirm 1) that is the db to be tested? 2) that it is currently running?

Thanks

**Comment by dww100** [comment]
Try `admin` or `daveww`.
Seems to be authenticating fine for me using those databases

**Comment by vivek-bala** [comment]
Ok 'admin' worked. The script was successful when run within a conda env as well.

To be sure, I have ```database_url='mongodb://radical:fg*2GT3^eB@crick.chem.ucl.ac.uk:27017/admin'``` in the script (or the same value for RADICAL_PILOT_DBURL). 

Not sure if this makes a difference, but are you using bash shell or csh? (maybe the url is not being parsed correctly)

**Comment by dww100** [comment]
The original point was that we can run as your suggest - but only once we have setup the database and only when that database is the admin database against which users are authenticated.

The radical tools documentation states that we can create new databases on the fly by changing the database URL (whether directly input or from the environment variable).

We were hoping that individual users would have separate databases. Is this not the way things are supposed to work?

**Comment by vivek-bala** [comment]
> The original point was that we can run as your suggest - but only once we have setup the database and only when that database is the admin database against which users are authenticated.
We were hoping that individual users would have separate databases. Is this not the way things are supposed to work?

Ah yes. I lost track of the 'on the fly' part. Thanks for reminding.  Let me test if its fully supported on both entk and the rp layer.

**Comment by sjzasada** [comment]
I've independently tried to run the example script Dave is referring to and have found that the database is not created on the fly.

**Comment by vivek-bala** [comment]
Following up on the discussions in https://github.com/radical-cybertools/radical.pilot/issues/1242, it seems to be incorrectly mentioned in the documentation that a database can be created on the fly. Due to difficulties in automation of the user authentication and management, RP or EnTK might not be the right place. 

A separate user management layer over EnTK could be something we can think about. Thoughts?
